drivers/flash/disk_flash.c

```diff
@@ -0,0 +1,344 @@
+	uint16_t const sector_buf_size;
+	uint32_t disk_total_sector_cnt;
+	uint32_t disk_sector_size;
+};
+
+struct disk_flash_config {
+	uint32_t flash_size;
+	struct flash_parameters flash_parameters;
+	char *disk_name;
+	uint32_t disk_offset;
+
+	IF_ENABLED(CONFIG_FLASH_PAGE_LAYOUT, (struct flash_pages_layout fpl;))
+};
+
+static void acquire_device(const struct device *dev)
+{
+	struct disk_flash_data *const data = dev->data;
+
+	k_sem_take(&data->sem, K_FOREVER);
+}
+
+static void release_device(const struct device *dev)
+{
+	struct disk_flash_data *const data = dev->data;
+
+	k_sem_give(&data->sem);
+}
+
+static int check_boundary(const struct device *dev, off_t offset, size_t len)
+{
+	int ret = 0;
+	const struct disk_flash_config *cfg = dev->config;
+	size_t operation_end = offset + len;
+
+	if (cfg->flash_size < operation_end || offset < 0) {
+		LOG_ERR("Operation out of bounds");
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static int disk_flash_read(const struct device *dev, off_t offset, void *dst, size_t len)
+{
+	int ret = 0;
+	uint8_t *dst_buf = dst;
+	const struct disk_flash_config *cfg = dev->config;
+	struct disk_flash_data *data = dev->data;
+	uint32_t write_cursor = 0;
+
+	acquire_device(dev);
+
+	ret = check_boundary(dev, offset, len);
+	if (ret != 0) {
+		goto release_device;
+	}
+
+	if (data->disk_sector_size == 0) {
+		ret = -EINVAL;
+		goto release_device;
+	}
+
+	while (write_cursor < len) {
+		uint32_t read_cursor = cfg->disk_offset + offset + write_cursor;
+		uint32_t read_sector = read_cursor / data->disk_sector_size;
+		uint32_t read_padding = read_cursor % data->disk_sector_size;
+		uint32_t len_write = (len - write_cursor);
+
+		if (read_sector < 0 || read_sector >= data->disk_total_sector_cnt) {
+			ret = -EINVAL;
+			goto release_device;
+		}
+
+		if (disk_access_read(cfg->disk_name, data->sector_buf, read_sector, 1)) {
+			ret = -EINVAL;
+			goto release_device;
+		}
+
+		if (len_write > data->sector_buf_size - read_padding) {
+			len_write = data->sector_buf_size - read_padding;
+		}
+
+		memcpy(&dst_buf[write_cursor], &data->sector_buf[read_padding], len_write);
+		write_cursor += len_write;
+	}
+
+release_device:
+	release_device(dev);
+	return ret;
+}
+
+static int disk_flash_write(const struct device *dev, off_t offset, const void *src, size_t len)
+{
+	int ret = 0;
+	const uint8_t *src_buf = src;
+	const struct disk_flash_config *cfg = dev->config;
+	struct disk_flash_data *data = dev->data;
+	uint32_t read_cursor = 0;
+
+	acquire_device(dev);
+
+	ret = check_boundary(dev, offset, len);
+	if (ret != 0) {
+		goto release_device;
+	}
+
+	if (data->disk_sector_size == 0) {
+		ret = -EINVAL;
+		goto release_device;
+	}
+
+	while (read_cursor < len) {
+		uint32_t write_cursor = cfg->disk_offset + offset + read_cursor;
+		uint32_t write_sector = write_cursor / data->disk_sector_size;
+		uint32_t write_padding = write_cursor % data->disk_sector_size;
+		uint32_t len_write = (len - read_cursor);
```

Comment:
I do not think it is very efficient to keep recalculating these in loop.

Reply:
@de-nordic 
I agree that this is not very efficient but I don't see a way around it since the driver allows arbitrary write/read sizes and does not need to be sector aligned with the disk.

I understood you believe, that read/write/erase granularity must be equal to the underlying disk device. The idea of this flash driver is to offer a read/write/erase granularity of a single byte on top of a disk with arbitrary sector sizes. So operations typically:
-read sector,
-update sector,
-write sector,
-loop

Reply:
But unless done this way, how would you simulate a flash with potentially different properties then underlying device?

Would you prefer, that `flash_disk` inherits properties from underlying device? If so, it would probably no longer be suitable for our use-case.
I see the following possibilities:
1. be at maximum permissive granularity (1 byte, current proposal)
2. be user configurable (with Kconfig?)
3. same as underlying device (severely limiting the usefulness of it to emulate a device with different properties on top of another)

Of course with 1. it is expected, that caller calls with big junks also. With 3. it is not possible to use as secondary partition for `mcuboot` if underlying device has different properties to main flash. Which would nullify the main benefit of the `flash_disk` module.

Or am I missing something? Can primary and secondary partition for mcuboot have different properties?
If yes, then of course, the story is entirely different.

Thanks for the quick responses and instructing reviews.

Reply:
Or did I completely miss your point and you suggestions is just to improve simply the code to:
- handle initial (potentially misaligned) sector
- handle most sectors with single call to `disk_access
- handle final (potentially misaligned) sector

Reply:
If I understood correctly this is exactly the Raison d'être of `flash_disk`. It provides a non-block-device interface to a part of a block device.

Reply:
The problem is that even flash drivers have something that is called "write-block-size" for ea reason. If you try to write size below that value you will end up damaging flash faster than expected by endurance characteristic of a device.
The problem we have here is that, for example, MCUboot will try to write some flags and it will do at 8 byte pace, which means that each time this happens the sector will be rewritten.
In case when MCUboot writes down status of swap operation it will write mark for each page it managed to swap, which means that the sector will be re-written as many times as the status is updated.
I am not sure how to solve this, but I do not think that locking the driver with 1 byte alignment is good idea.
If somebody picks the driver and starts to write info to SD card by doing writes by that alignment, the person will damage the SD card in unexpectedly short time - which means that we then have Zephyr issue logged.

Reply:
@de-nordic Why would/does McuBoot use a 8byte pace when it could write any size possible?
Anyway in order for MCUBoot to work proberly the flash page size needs to be the same as the page size of the primary flash partition. Therefore I have added a DTS property again to be able to adjust this freely. The code still can handle any size and will write as much bytes as possible at once in order to not wear out the underlying disk.

Reply:
> @de-nordic Why would/does McuBoot use a 8byte pace when it could write any size possible? Anyway in order for MCUBoot to work proberly the flash page size needs to be the same as the page size of the primary flash partition. Therefore I have added a DTS property again to be able to adjust this freely. The code still can handle any size and will write as much bytes as possible at once in order to not wear out the underlying disk.

Looks like design requires 4:
https://github.com/mcu-tools/mcuboot/blob/c43a20fd193dc445faa9b7f0d2ed77759a6aca8f/docs/design.md?plain=1#L468-L474

Zephyr uses at least 8:
https://github.com/mcu-tools/mcuboot/blob/c43a20fd193dc445faa9b7f0d2ed77759a6aca8f/boot/bootutil/include/bootutil/bootutil_public.h#L88-L91

sure about the flash page size?
https://github.com/mcu-tools/mcuboot/blob/c43a20fd193dc445faa9b7f0d2ed77759a6aca8f/docs/design.md?plain=1#L184-L195

Reply:
@de-nordic 

The primary and the secondary image need the same amount of sectors otherwise mcuboot fails at startup with the error message:
`W: Cannot upgrade: not a compatible amount of sectors`
But we are not using the scratch partition instead we are using the `Swap without using scratch`<

As far as I understood the problem is the writing of the partition header which needs to be written with every sector that is swapped. As I understood it there is nothing that can be done about this. But wouldn't the same problem occur with a flash since only a whole page can be written? 

BTW
I adjusted the driver in order to better emulate the write/erase behaviour by using a bitwise AND. I also added the write-block-size as a dts attribute which per default gets set to the sector size of the underlying this. This way the user explicitly has to set a lower block-size himself hopefully knowing the implication on sd-card-lifetime.

Reply:
Thank you for elaborating your concerns, I think, that I understand better now. Did I understand your concerns correclty, that:
1. The trailer containing meta-information about the slot is written at the end of each slot. So there is a trailer at the end of the secondary slot on the flash_disk. Did I understand correctly, that this would mean, that whereas on a flash the trailer is erased once and then progressively written the flash-disk would need to read-update-write the sector containing the trailer with each status update which concerns you because of too much wear on disks without wear-leveling?
2. You are worried, that the write of a sector is not an atomic operation, such that it might end up being half-written for example in case of power-loss? Leading to possible damage to the system.

> There is another issue that there is very small chance that you can break the status update when record is added, but when you break it while write block happens, what will happen is that there will not longer be record of any swap happening and, at the same time, you will have two damaged slots (partially swapped).

Are you refering to the fact, that swapping might fail when using `swap-using-move` on an image that is too big (100% of available space) because it can not move down the primary imgage block-wise when there is no free block? Whereas this can not happen when using `swap-using-scratch`? Or is there another reason for recommending `swap using scratch`?

Thanks

Reply:
Both with `swap-using-move` and `swap-using-scratch` the result of a block move is written to flash. When using a disk and the write of the last sector fails you could loose all information about the process (all information is on one disk sector), while on flash only the last block move information might be incorrect. If all information is lost there is no way to continue the swap and you are left with an unrecoverable system.

Reply:
Yes, the strategy does not matter.

Reply:
@de-nordic @Laczen 

Shouldn't the danger of sector write fail be lowered since when not written entirely the sd-card should still contain a partially correct sector? Or how is the status of the swap tracked? By settings bytes to 0?

Also I do not currently see any mitigation for the two problems brought up in this thread. Neither for the problem with the high disk usage due to sector overwrites nor for the problem with partial writes. Do you see any mitigation possibilites here or what would/should be the way forward?